opencxl/cxl/component/cxl_memory_hub.py

```diff
@@ -79,45 +79,79 @@ def __init__(self, config: CxlComplexHostConfig):
-        )
-        self._host_simple_processor = HostLlcIoGen(host_processor_config)
+    def get_memory_ranges(self):
+        return self._cache_controller.get_memory_ranges()
 
-        self._dev_mmio_ranges: list[tuple[int, int]] = []
-        self._dev_mem_ranges: list[tuple[int, int]] = []
+    def add_mem_range(self, addr, size, addr_type: MEM_ADDR_TYPE):
+        self._cache_controller.add_mem_range(addr, size, addr_type)
 
-    def append_dev_mmio_range(self, base, size):
-        self._dev_mmio_ranges.append((base, size))
+    def _cfg_addr_to_bdf(self, cfg_addr):
+        return create_bdf(
+            (cfg_addr >> 20) & 0xFF,  # bus bits, n = 8
+            (cfg_addr >> 15) & 0x1F,
+            (cfg_addr >> 12) & 0x07,
+        )
 
-    def append_dev_mem_range(self, base, size):
-        self._dev_mem_ranges.append((base, size))
+    async def _send_mem_request(self, packet: MemoryRequest) -> MemoryResponse:
+        await self._processor_to_cache_fifo.request.put(packet)
+        resp = await self._processor_to_cache_fifo.response.get()
+        assert resp.status == MEMORY_RESPONSE_STATUS.OK
+        return resp
+
+    async def load(self, addr: int, size: int) -> int:
+        addr_type = self._cache_controller.get_mem_addr_type(addr)
+        match addr_type:
+            case MEM_ADDR_TYPE.DRAM | MEM_ADDR_TYPE.CXL_CACHED | MEM_ADDR_TYPE.CXL_CACHED_BI:
+                packet = MemoryRequest(MEMORY_REQUEST_TYPE.READ, addr, size)
+                resp = await self._send_mem_request(packet)
+                return resp.data
+            case MEM_ADDR_TYPE.CXL_UNCACHED:
+                packet = MemoryRequest(MEMORY_REQUEST_TYPE.UNCACHED_READ, addr, size)
+                resp = await self._send_mem_request(packet)
+                return resp.data
+            case MEM_ADDR_TYPE.MMIO:
+                return await self._root_complex.read_mmio(addr, size)
+            case MEM_ADDR_TYPE.CFG:
+                bdf = self._cfg_addr_to_bdf(addr)
+                offset = addr & 0xFFF
+                return await self._root_complex.read_config(bdf, offset, size)
+            case _:
+                raise Exception(self._create_message(f"Address 0x{addr:x} is OOB."))
+
+    async def store(self, addr: int, size: int, data: int):
+        addr_type = self._cache_controller.get_mem_addr_type(addr)
+        match addr_type:
+            case MEM_ADDR_TYPE.DRAM | MEM_ADDR_TYPE.CXL_CACHED | MEM_ADDR_TYPE.CXL_CACHED_BI:
+                packet = MemoryRequest(MEMORY_REQUEST_TYPE.WRITE, addr, size, data)
+                await self._send_mem_request(packet)
+            case MEM_ADDR_TYPE.CXL_UNCACHED:
+                packet = MemoryRequest(MEMORY_REQUEST_TYPE.UNCACHED_WRITE, addr, size, data)
+                await self._send_mem_request(packet)
+            case MEM_ADDR_TYPE.MMIO:
+                await self._root_complex.write_mmio(addr, size, data)
+            case MEM_ADDR_TYPE.CFG:
+                bdf = self._cfg_addr_to_bdf(addr)
```

Comment:
Does this assume CFG address always start from 0?
Shouldn't the address be subtracted by the base address of CFG base address if CFG base address is not 0?

Reply:
Fixed